File: cpp/rsagen.cpp
#include <openssl/bn.h>
#include <openssl/rand.h>
#include <openssl/pem.h>
#include <openssl/rsa.h>
#include <iostream>
#include <fstream>
#include <cstring>
#include <stdexcept>

struct keyInfo
{
    BIGNUM *d;
    BIGNUM *p;
    BIGNUM *q;
    BIGNUM *yn;
    BIGNUM *e;
    BIGNUM *n;
    BIGNUM *dmp1;
    BIGNUM *dmq1;
    BIGNUM *iqmp1;
};

// generate prime value using input from stdin
BIGNUM *genPrime(int valSize, BN_CTX *bnCtx)
{
    BIGNUM *value = BN_new();
    BIGNUM *two = BN_new();
    BN_set_word(two, 2);

    unsigned char buffer[valSize / 8];
    int isPrime = 0;

    std::cin.read(reinterpret_cast<char *>(buffer), valSize / 8);

    if (!std::cin)
    {
        std::cerr << "Failed to read random bytes from stdin." << std::endl;
        return nullptr;
    }

    buffer[valSize / 8 - 1] |= 0x01; // to make sure the value is odd

    BN_bin2bn(buffer, valSize / 8, value);

    while (!isPrime)
    {
        BN_add(value, value, two);
        isPrime = BN_is_prime_fasttest(value, 128, NULL, bnCtx, NULL, 1);   
    }

    BN_free(two);

    return value;
}

// generate RSA key-pair values
void rsaKeyGen(keyInfo *key, unsigned long exponent, int keySize)
{
    // choose two large prime numbers
    BN_CTX *ctx = BN_CTX_new();
    BIGNUM *p = genPrime(keySize / 2, ctx);
    BIGNUM *q = genPrime(keySize / 2, ctx);

    // n = pq
    BIGNUM *n = BN_new();
    BN_mul(n, p, q, ctx);

    // compute Carmichael's totient function
    BIGNUM *gcd = BN_new();
    BIGNUM *yn = BN_new();
    BIGNUM *pm = BN_new();
    BIGNUM *qm = BN_new();
    BN_sub(pm, p, BN_value_one());
    BN_sub(qm, q, BN_value_one());
    BN_gcd(gcd, pm, qm, ctx);
    BN_mul(yn, pm, qm, ctx);
    BN_div(yn, NULL, yn, gcd, ctx);

    // exponent
    BIGNUM *e = BN_new();
    BN_set_word(e, exponent);

    // check gcd(e, λ(n)) = 1
    BIGNUM *g = BN_new();
    BN_gcd(g, e, yn, ctx);
    
    if (BN_is_one(g) != 1)
    {
        throw std::logic_error("check gcd(e, λ(n)) != 1");
    }

    // modular inverse
    BIGNUM *d = BN_new();
    BN_mod_inverse(d, e, yn, ctx);

    // Chinese remainder theorem related constants
    BIGNUM *dmp1 = BN_new();
    BIGNUM *dmq1 = BN_new();
    BIGNUM *iqmp = BN_new();
    BIGNUM *dm = BN_new();

    BN_sub(dm, d, BN_value_one());
    BN_mod(dmp1, d, pm, ctx);
    BN_mod(dmq1, d, qm, ctx);
    BN_mod_inverse(iqmp, q, p, ctx);

    // results
    key->d = d;
    key->e = e;
    key->p = p;
    key->q = q;
    key->yn = yn;
    key->n = n;
    key->dmp1 = dmp1;
    key->dmq1 = dmq1;
    key->iqmp1 = iqmp;

    // clean
    BN_free(gcd);
    BN_free(g);
}

// simple test for small numbers 
// (not used for RSA key-pair generation)
bool isPrime(unsigned long n)
{
    if (n == 2 || n == 3)
        return true;

    if (n <= 1 || n % 2 == 0 || n % 3 == 0)
        return false;

    for (int i = 5; i * i <= n; i += 6)
    {
        if (n % i == 0 || n % (i + 2) == 0)
            return false;
    }

    return true;
}

int main(int argc, char const *argv[])
{
    unsigned long exponent = 65537; // 2^16+1
    int keySize = 2048;

    // command line arguments
    if (argc < 3)
    {
        std::cerr << "Usage: " << argv[0] << " <private_key_file> <public_key_file> [-s | -e]" << std::endl
                  << "-e: set exponent value (65535 default)\n-s: key size (2048 default)" << std::endl;
        return 1;
    }
    else if (argc > 3)
    {
        for (int i = 3; i < argc; i++)
        {
            if (strcmp(argv[i], "-e") == 0)
            {
                exponent = std::stoul(argv[i + 1]);
                if (exponent < 3 || !isPrime(exponent))
                {
                    std::cerr << "Error: Invalid exponent value" << std::endl;
                }
            }
            else if (strcmp(argv[i], "-s") == 0)
            {
                if (std::atoi(argv[i + 1]) % 8 != 0)
                {
                    std::cerr << "Error: Make sure that the key size is a multiple of 8" << std::endl;
                    return 1;
                }
                keySize = std::atoi(argv[i + 1]);
            }
        }
    }

    const char *privFile = argv[1];
    const char *pubFile = argv[2];

    // generate key-pair
    keyInfo keyPair;
    while(1)
    {
        try
        {
            rsaKeyGen(&keyPair, exponent, keySize);
            break;
        } catch(std::logic_error& e)
        {
            continue;
        }
    }
    
    // save keys in PEM format
    RSA *rsa = RSA_new();
    RSA_set0_key(rsa, keyPair.n, keyPair.e, keyPair.d);
    RSA_set0_factors(rsa, keyPair.p, keyPair.q);
    RSA_set0_crt_params(rsa, keyPair.dmp1, keyPair.dmq1, keyPair.iqmp1);

    FILE *pubFp = fopen(pubFile, "w");
    PEM_write_RSAPublicKey(pubFp, rsa);
    fclose(pubFp);

    FILE *privFp = fopen(privFile, "w");
    PEM_write_RSAPrivateKey(privFp, rsa, 0, 0, 0, 0, 0);

    RSA_free(rsa);

    return 0;
}

File: go/rsagen.go
package main

import (
	"crypto/rsa"
	"crypto/x509"
	"encoding/pem"
	"errors"
	"fmt"
	"math/big"
	"os"
	"strconv"
)

type keyInfo struct {
	d     *big.Int
	p     *big.Int
	q     *big.Int
	yn    *big.Int
	e     *big.Int
	n     *big.Int
	dmp1  *big.Int
	dmq1  *big.Int
	iqmp1 *big.Int
}

func genPrime(valSize int) (*big.Int, error) {
	data := make([]byte, valSize/8)

	_, err := os.Stdin.Read(data)

	if err != nil {
		return nil, err
	}

	two := new(big.Int).SetInt64(2)

	data[valSize/8-1] |= 0x01
	value := new(big.Int).SetBytes(data[:valSize/8])

	isPrime := false

	for !isPrime {
		value.Add(value, two)
		isPrime = value.ProbablyPrime(128)
	}

	return value, nil
}

func rsaKeyGen(keySize int, exponent uint64) (*keyInfo, error) {
	// Const
	one := new(big.Int).SetInt64(1)

	// generate two primes, p and q
	p, _ := genPrime(keySize / 2)
	q, _ := genPrime(keySize / 2)

	// n = pq
	n := new(big.Int)
	n.Mul(p, q)

	// compute Carmichael's totient function
	gcd := new(big.Int)
	yn := new(big.Int)
	pm := new(big.Int)
	qm := new(big.Int)
	pm.Sub(p, one)
	qm.Sub(q, one)
	gcd.GCD(nil, nil, pm, qm)
	yn.Mul(pm, qm)
	yn.Div(yn, gcd)

	// exponent
	e := new(big.Int)
	e.SetUint64(exponent)

	// check gcd(e, λ(n)) = 1
	gcd.GCD(nil, nil, e, yn)
	if gcd.Cmp(one) != 0 {
		return nil, errors.New("check gcd(e, λ(n)) != 1")
	}

	// modular inverse
	d := new(big.Int)
	d.ModInverse(e, yn)

	// Chinese remainder theorem related constants
	dmp1 := new(big.Int)
	dmq1 := new(big.Int)
	iqmp := new(big.Int)
	dm := new(big.Int)

	dm.Sub(d, one)
	dmp1.Sub(d, pm)
	dmq1.Sub(d, qm)
	iqmp.ModInverse(q, p)

	key := keyInfo{d, p, q, yn, e, n, dmp1, dmq1, iqmp}

	return &key, nil
}

// simple test for small numbers
// (not used for RSA key-pair generation)
func isPrime(n uint64) bool {
	if n <= 1 {
		return false
	}
	for i := uint64(2); i*i <= n; i++ {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func main() {
	keySize := 2048
	exponent := 65537
	var err error

	if len(os.Args) < 3 {
		fmt.Printf("Usage: %s <private_key_file> <public_key_file> [-s | -e]\n", os.Args[0])
		fmt.Println("-e: set exponent value (default:", exponent, ")\n-s: key size (default:", keySize, ")")
		os.Exit(1)
	} else if len(os.Args) > 3 {
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "-e" {
				exponentArg := os.Args[i+1]
				if exponent, err = strconv.Atoi(exponentArg); err != nil || exponent < 3 || !isPrime(uint64(exponent)) {
					fmt.Println("Error: Invalid exponent value")
					os.Exit(1)
				}
			} else if os.Args[i] == "-s" {
				keySizeArg := os.Args[i+1]

				if keySize, err = strconv.Atoi(keySizeArg); err != nil || keySize%8 != 0 {
					fmt.Println("Error: Invalid key size")
					os.Exit(1)
				}
			}
		}
	}

	var keyVals *keyInfo

	for true {
		keyVals, err = rsaKeyGen(keySize, uint64(exponent))
		if err == nil {
			break
		}
	}

	keyPair := &rsa.PrivateKey{
		PublicKey: rsa.PublicKey{
			N: keyVals.n,
			E: exponent,
		},
		D:      keyVals.d,
		Primes: []*big.Int{keyVals.p, keyVals.q},
	}

	// Private key to PEM
	privateKeyPEM := &pem.Block{
		Type:  "RSA PRIVATE KEY",
		Bytes: x509.MarshalPKCS1PrivateKey(keyPair),
	}

	privateKeyFile, err := os.Create(os.Args[1])
	if err != nil {
		fmt.Println("Error creating private key file:", err)
		return
	}
	defer privateKeyFile.Close()

	err = pem.Encode(privateKeyFile, privateKeyPEM)
	if err != nil {
		fmt.Println("Error encoding private key to PEM:", err)
		return
	}

	// Public key to PEM
	publicKeyPEM, err := x509.MarshalPKIXPublicKey(&keyPair.PublicKey)
	if err != nil {
		fmt.Println("Error marshaling public key:", err)
		return
	}

	publicKeyFile, err := os.Create(os.Args[2])
	if err != nil {
		fmt.Println("Error creating public key file:", err)
		return
	}
	defer publicKeyFile.Close()

	err = pem.Encode(publicKeyFile, &pem.Block{
		Type:  "PUBLIC KEY",
		Bytes: publicKeyPEM,
	})
	if err != nil {
		fmt.Println("Error encoding public key to PEM:", err)
		return
	}
}
